internal/prefixcollector/collector_test.go

```diff
@@ -0,0 +1,221 @@
+
+package prefixcollector_test
+
+import (
+	"cmd-exclude-prefixes-k8s/internal/prefixcollector"
+	"context"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/fsnotify/fsnotify"
+	"github.com/ghodss/yaml"
+	"github.com/stretchr/testify/require"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes/fake"
+
+	"cmd-exclude-prefixes-k8s/internal/utils"
+	"io/ioutil"
+	"testing"
+)
+
+const (
+	configMapPath            = "./testfiles/configMap.yaml"
+	kubeConfigMapPath        = "./testfiles/kubeAdmConfigMap.yaml"
+	testExcludedPrefixesPath = "./testfiles/excludedPrefixes.yaml"
+	configMapName            = "test"
+	configMapNamespace       = "default"
+)
+
+type dummyPrefixSource struct {
+	prefixes []string
+}
+
+func (d *dummyPrefixSource) Prefixes() []string {
+	return d.prefixes
+}
+
+func newDummyPrefixSource(prefixes []string, cond *sync.Cond) *dummyPrefixSource {
+	go func() {
+		for {
+			cond.Broadcast()
+			<-time.After(time.Second)
+		}
+	}()
+	return &dummyPrefixSource{prefixes}
+}
+
+func TestCollectorWithDummySources(t *testing.T) {
+	cond := sync.NewCond(&sync.Mutex{})
+	sources := []prefixcollector.ExcludePrefixSource{
+		newDummyPrefixSource(
+			[]string{
+				"127.0.0.1/16",
+				"127.0.2.1/16",
+				"168.92.0.1/24",
+			},
+			cond,
+		),
+		newDummyPrefixSource(
+			[]string{
+				"127.0.3.1/18",
+				"134.56.0.1/8",
+				"168.92.0.1/16",
+			},
+			cond,
+		),
+	}
+
+	expectedResult := []string{
+		"127.0.0.0/16",
+		"168.92.0.0/16",
+		"134.0.0.0/8",
+	}
+
+	testCollector(t, cond, expectedResult, sources)
+}
+
+func TestKubeAdmConfigSource(t *testing.T) {
+	expectedResult := []string{
+		"10.244.0.0/16",
+		"10.96.0.0/12",
+	}
+
+	cond := sync.NewCond(&sync.Mutex{})
+	ctx := createConfigMap(t, prefixcollector.KubeNamespace, kubeConfigMapPath)
+	sources := []prefixcollector.ExcludePrefixSource{
+		prefixcollector.NewKubeAdmPrefixSource(ctx, cond),
+	}
+
+	testCollector(t, cond, expectedResult, sources)
+}
+
+func TestConfigMapSource(t *testing.T) {
+	expectedResult := []string{
+		"168.0.0.0/10",
+		"1.0.0.0/11",
+	}
+	cond := sync.NewCond(&sync.Mutex{})
+
+	ctx := createConfigMap(t, configMapNamespace, configMapPath)
+	sources := []prefixcollector.ExcludePrefixSource{
+		prefixcollector.NewConfigMapPrefixSource(ctx,
+			cond,
+			configMapName,
+			configMapNamespace,
+		),
+	}
+
+	testCollector(t, cond, expectedResult, sources)
+}
+
+func createConfigMap(t *testing.T, namespace, configPath string) context.Context {
+	ctx := context.Background()
+	clientSet := fake.NewSimpleClientset()
+	configMap := getConfigMap(t, configPath)
+
+	ctx = context.WithValue(ctx, prefixcollector.ClientSetKey, clientSet)
+	_, _ = clientSet.CoreV1().
+		ConfigMaps(namespace).
+		Create(ctx, configMap, metav1.CreateOptions{})
+
+	return ctx
+}
+
+func testCollector(t *testing.T, cond *sync.Cond, expectedResult []string, sources []prefixcollector.ExcludePrefixSource) {
```

Comment:
Could you also add `defer goleak.VerifyNone(t, goleak.IgnoreCurrent())` check in each test?

Reply:
Note: You'll need to add dep go.uber.org/goleak

Reply:
Added.